app/src/bat_charger.cpp

```diff
@@ -22,13 +22,73 @@ LOG_MODULE_REGISTER(bat_charger, CONFIG_BAT_LOG_LEVEL);
 extern DeviceStatus dev_stat;
 extern LoadOutput load;
 
+uint32_t milli_seconds_in_float = 0;
+uint32_t float_reset_duration = 600000;             // 10 minutes in milliseconds
+const uint32_t soc_scaled_hundred_percent = 100000; // 100% charge = 100000
+const uint32_t soc_scaled_max =
+    2 * soc_scaled_hundred_percent; // allow soc to track up higher than 100% to gauge efficiency
+
 // DISCHARGE_CURRENT_MAX used to estimate current-compensation of load disconnect voltage
 #if BOARD_HAS_LOAD_OUTPUT
 #define DISCHARGE_CURRENT_MAX DT_PROP(DT_CHILD(DT_PATH(outputs), load), current_max)
 #else
 #define DISCHARGE_CURRENT_MAX DT_PROP(DT_PATH(pcb), dcdc_current_max)
 #endif
 
+float calculate_initial_soc(float battery_voltage_mV)
+{
+    // TODO will need to add 24 V compatability
+    const uint32_t soc_scaled_hundred_percent = 100000;
+    const uint8_t voltages_size = 10;
+    const float batt_soc_voltages[voltages_size] = { 12720, 12600, 12480, 12360, 12240,
+                                                     12120, 12000, 11880, 11760, 11640 };
+
+    uint8_t index;
+    for (index = 0; index < voltages_size; index++) {
+        if (battery_voltage_mV > batt_soc_voltages[index]) {
+            break;
+        }
+    }
+    return (voltages_size - index) * (soc_scaled_hundred_percent / voltages_size);
```

Comment:
We should better use linear interpolation and not use the nearest value. There is a function in the [helper.cpp of the BMS firmware](https://github.com/LibreSolar/bms-firmware/blob/main/app/src/helper.cpp) which could be used.